File: Javascript/traversy-js-challenges-main/09-sorting-algorithms/09-merge-sort-implementation/merge-sort.js
function merge(left, right) {
    console.log('************************************************************************')
    console.log('------------------Merge----------------------')
    console.log(`Now merging left (${left}) with right ${right}`)
    const merged = [];
    let leftIndex = 0;
    let rightIndex = 0;
  
    while (leftIndex < left.length && rightIndex < right.length) {
      if (left[leftIndex] < right[rightIndex]) {
        merged.push(left[leftIndex]);
        leftIndex++;
      } else {
        merged.push(right[rightIndex]);
        rightIndex++;
      }
    }

    console.log(`After sorting: ${merged}`)

    const remainingLeft = left.slice(leftIndex);

    console.log(`Remaining left: ${remainingLeft}`)
;
    const remainingRight = right.slice(rightIndex);
  
    console.log(`Remaining right: ${remainingRight}`)

    const mergeOutput =  merged.concat(remainingLeft).concat(remainingRight);

    console.log(`Merged output: ${mergeOutput}`)
    console.log('************************************************************************')

    return mergeOutput;
  }

  function mergeSort(arr) {
    console.log('<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>');
    console.log(`Recursively calling mergeSort: ${arr}`);
    if (arr.length <= 1) {
        console.log(`@@@@@@@Only one item in array -- return (unwind)@@@@@@@@@@`)
      return arr;
    }
  
    const mid = Math.floor(arr.length / 2);
    console.log('++++++++++++++++++++++++Dividing the array into two+++++++++++++++++')
    const left = mergeSort(arr.slice(0, mid));
    const right = mergeSort(arr.slice(mid));
    console.log('+++++++++++++++++++Now merging two halves+++++++++++++++++++++++')
    return merge(left, right);
  }

module.exports = mergeSort;

File: Python/SortingAlgorithms/sorting_functions/merge_sort.py

def merge(left_list, right_list):
    left_index = 0
    right_index = 0
    sorted_items = []
    while left_index < len(left_list) and right_index < len(right_list):
        if left_list[left_index] < right_list[right_index]:
            sorted_items.append(left_list[left_index])
            left_index += 1
        else:
            sorted_items.append(right_list[right_index])
            right_index += 1
    return sorted_items + left_list[left_index:] + right_list[right_index:]


def merge_sort(list_to_be_sorted):
    if len(list_to_be_sorted) <= 1:
        return list_to_be_sorted
    mid_point = len(list_to_be_sorted) // 2
    left_list = merge_sort(list_to_be_sorted[0:mid_point])
    right_list = merge_sort(list_to_be_sorted[mid_point:])
    return merge(left_list, right_list)
